File: 415-AddStrings/415-AddStrings.java
/*

415. Add Strings
Easy

Given two non-negative integers num1 and num2 represented as string, return the sum of num1 and num2.

Note:

    The length of both num1 and num2 is < 5100.
    Both num1 and num2 contains only digits 0-9.
    Both num1 and num2 does not contain any leading zero.
    You must not use any built-in BigInteger library or convert the inputs to integer directly.

*/

class Solution {
    public String addStrings(String num1, String num2) {
        Stack<Integer> s1 = new Stack<>();
        Stack<Integer> s2 = new Stack<>();
        for(int i =0; i<num1.length(); i++)
            s1.push(Character.getNumericValue(num1.charAt(i)));
        for(int i =0; i<num2.length(); i++)
            s2.push(Character.getNumericValue(num2.charAt(i)));
        int carry = 0;
        StringBuilder sb = new StringBuilder();
        while(!s1.isEmpty() || !s2.isEmpty() || carry != 0){
            int n1 = s1.isEmpty() ? 0 : s1.pop();
            int n2 = s2.isEmpty() ? 0 : s2.pop();
            int val = n1 + n2 + carry;
            if (val >= 10){
                val = val % 10;
                carry = 1;
            }
            else
                carry = 0;
            
            sb.append(val);
        }
        return sb.reverse().toString();
    }
}

File: 415-AddStrings/415-AddStrings.py
"""

415. Add Strings

Given two non-negative integers, num1 and num2 represented as string, return the sum of num1 and num2 as a string.

You must solve the problem without using any built-in library for handling large integers (such as BigInteger). You must also not convert the inputs to integers directly.

Example 1:

Input: num1 = "11", num2 = "123"
Output: "134"

Example 2:

Input: num1 = "456", num2 = "77"
Output: "533"

Example 3:

Input: num1 = "0", num2 = "0"
Output: "0"

Constraints:

    1 <= num1.length, num2.length <= 104
    num1 and num2 consist of only digits.
    num1 and num2 don't have any leading zeros except for the zero itself.

"""

class Solution:
    def addStrings(self, num1: str, num2: str) -> str:
        i, j, carry = len(num1) - 1, len(num2) - 1, 0
        _sum = 0
        q = deque()
        while i >= 0 or j >= 0:
            first = num1[i] if i >= 0 else 0
            sec = num2[j] if j >= 0 else 0
            _sum = int(first) + int(sec) + carry
            carry = _sum // 10
            _sum = _sum % 10
            q.appendleft(str(_sum))
            i -= 1
            j -= 1
        if carry == 1: q.appendleft('1')
        return "".join(list(q))
